smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/provider/UserTuneProvider.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.usertune.provider;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.parsing.SmackParsingException;
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParser.Event;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+import org.jivesoftware.smackx.usertune.element.UserTuneElement;
+
+/**
+ * This is the Provider Class for {@link UserTuneElement}.
+ */
+public class UserTuneProvider extends ExtensionElementProvider<UserTuneElement> {
+
+    private static final Logger LOGGER = Logger.getLogger(UserTuneProvider.class.getName());
+    public static final UserTuneProvider INSTANCE = new UserTuneProvider();
+
+    @Override
+    public UserTuneElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment)
+            throws XmlPullParserException, IOException, SmackParsingException {
+
+        UserTuneElement.Builder builder = UserTuneElement.getBuilder();
+        outerloop: while (true) {
+            XmlPullParser.Event tag = parser.next();
+            String name = parser.getName();
+            String namespace = parser.getNamespace();
+
+            switch (tag) {
+            case START_ELEMENT:
+                if (!UserTuneElement.NAMESPACE.equals(namespace)) {
+                    LOGGER.log(Level.FINE, "Foreign namespace " + namespace + " detected.");
+                    continue outerloop;
+                }
+                try {
+                    while (true) {
+                        switch (name) {
+                        case "artist":
+                            builder.setArtist(parser.nextText());
+                            break;
+                        case "length":
+                            builder.setLength(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "rating":
+                            builder.setRating(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "source":
+                            builder.setSource(parser.nextText());
+                            break;
+                        case "title":
+                            builder.setTitle(parser.nextText());
+                            break;
+                        case "track":
+                            builder.setTrack(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "uri":
+                            URI uri = new URI(parser.nextText());
```

Comment:
If you use `ParserUtils.getUriFromNextText()` here, then you can remove the whole try-catch block.